edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/ModuleIdentityLifecycleManager.cs

```diff
@@ -28,6 +28,18 @@ public ModuleIdentityLifecycleManager(IIdentityManager identityManager, ModuleId
 
         public async Task<IImmutableDictionary<string, IModuleIdentity>> GetModuleIdentitiesAsync(ModuleSet desired, ModuleSet current)
         {
+            IImmutableDictionary<string, Identity> identities = (await this.identityManager.GetIdentities()).ToImmutableDictionary(i => i.ModuleId);
+
+            // Need to remove any identities (except EA/EH) that are managed by EA but don't have a tracked module for in the ModuleSet.
+            IEnumerable<string> removeCurrentIdentities = identities.Where(
+                i => !(Constants.EdgeAgentModuleIdentityName.Equals(i.Key, StringComparison.Ordinal) || Constants.EdgeHubModuleIdentityName.Equals(i.Key, StringComparison.Ordinal))).Where(
+                i => Constants.ModuleIdentityEdgeManagedByValue.Equals(i.Value.ManagedBy, StringComparison.OrdinalIgnoreCase) &&
+                     !current.Modules.Any(m => ModuleIdentityHelper.GetModuleIdentityName(m.Key) == i.Key))
```

Comment:
You also need to compare against all the desired modules.

Reply:
Don't we only need to do that if there is an actual diff (non-empty?)

The flow I imagine we want (will let you or @ancaantochi correct me if I am wrong here :)) is
1. Get all the identities
2. Check what modules we currently have against the identities
3. If EA has missing module for an identity currently, then remove that identity
4. If there is a difference between desired and current, let normal workflow happen to create identities and deploy modules

Reply:
This cleanup logic will only remove identities for things that edgeAgent has (previously) seen, right?  Just want to ensure that host-level modules such as Device Update or Defender won't end up being removed accidentally.  They currently appear as regular module identities in the portal with "NA" status (by design, maybe not great design, but still intentional).